---
title: "Introduction to Single Cell RNAseq Part 3"
author: "UCD Bioinformatics Core"
output:
    html_document:
      keep_md: TRUE
---


Last Updated: July 15, 2022

# Part 3: Integrate multiple single cell samples


## Load libraries
```{r libraries, warning=FALSE,error=FALSE,message=FALSE}
library(Seurat)
```

## Load the Seurat object from the prior excercise and split to individual samples

```{r load_rdata, warning=FALSE,error=FALSE,message=FALSE}
load(file="pre_sample_corrected.RData")
experiment.aggregate
experiment.split <- SplitObject(experiment.aggregate, split.by = "ident")
```

## Normalize and find variable features for each individual sample

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}
experiment.split <- lapply(X = experiment.split, FUN=function(x){
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})
```

## Select features that are repeatedly variable across samples and find integration anchors

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}
features <- SelectIntegrationFeatures(object.list = experiment.split)
anchors <- FindIntegrationAnchors(object.list = experiment.split, anchor.features = features)
```


## Perform integration

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}
experiment.integrated <- IntegrateData(anchorset = anchors)
```


#### Question(s)

1. Explore the object "experiment.integrated" to see what information is available.

## PCA plot before integration

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}
experiment.test <- NormalizeData(object=experiment.integrated, assay="RNA")
experiment.test <- ScaleData(object=experiment.test, assay="RNA")
experiment.test <- FindVariableFeatures(object=experiment.test, assay="RNA")
experiment.test <- RunPCA(object=experiment.test, assay="RNA")
DimPlot(object = experiment.test, group.by="ident", reduction="pca", shuffle=TRUE)
```

## PCA plot after integration

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}
experiment.test <- ScaleData(object=experiment.integrated, assay="integrated")
experiment.test <- FindVariableFeatures(object=experiment.test, assay="integrated")
experiment.test <- RunPCA(object=experiment.test, assay="integrated")
DimPlot(object = experiment.test, group.by="ident", reduction="pca", shuffle=TRUE)
```

## Save the integrated data

```{r save, eval=TRUE}
save(experiment.integrated, file="sample_integrated.RData")
```

## Get the next Rmd file
```{r get_next_rmd, eval=FALSE}
download.file("https://raw.githubusercontent.com/ucdavis-bioinformatics-training/2022-March-Single-Cell-RNA-Seq-Analysis/main/data_analysis/scRNA_Workshop-PART4.Rmd", "scRNA_Workshop-PART4.Rmd")
```

## Session Information
```{r sessioninfo}
sessionInfo()
```
